lib/plugins/aws/deploy/lib/mergeIamTemplates.js

```diff
@@ -11,6 +11,47 @@ module.exports = {
   },
 
   merge() {
+    // resolve early if no functions are provided
+    if (!this.serverless.service.getAllFunctions().length) {
+      return BbPromise.resolve();
+    }
+
+    // resolve early if provider level role is provided
+    if ('role' in this.serverless.service.provider) {
+      return BbPromise.resolve();
+    }
+
+    // resolve early if all functions contain a custom role
+    const customRolesProvided = [];
+    this.serverless.service.getAllFunctions().forEach((functionName) => {
+      const functionObject = this.serverless.service.getFunction(functionName);
+      customRolesProvided.push('role' in functionObject);
+    });
+    if (_.isEqual(_.uniq(customRolesProvided), [true])) {
+      return BbPromise.resolve();
+    }
+
+    // merge in the iamRoleLambdaTemplate
+    const iamRoleLambdaExecutionTemplate = this.serverless.utils.readFileSync(
+      path.join(this.serverless.config.serverlessPath,
+        'plugins',
+        'aws',
+        'deploy',
+        'lib',
+        'iam-role-lambda-execution-template.json')
+    );
+    iamRoleLambdaExecutionTemplate.Properties.Path = this.provider.naming.getRolePath();
+    iamRoleLambdaExecutionTemplate.Properties.RoleName = this.provider.naming.getRoleName();
+    iamRoleLambdaExecutionTemplate.Properties.Policies[0]
+      .PolicyName = this.provider.naming.getPolicyName();
+
+    _.merge(
+      this.serverless.service.provider.compiledCloudFormationTemplate.Resources,
+      {
+        [this.provider.naming.getRoleLogicalId()]: iamRoleLambdaExecutionTemplate,
+      }
+    );
+
     this.serverless.service.getAllFunctions().forEach((functionName) => {
```

Comment:
This code block must be at the very beginning of the `merge()` method, otherwise bug #3174 will resurface. In other words, now that we *always* use log group resources, they must be merged regardless of whether we're using custom or default roles.